### ISSUE TITLE
[bug] semver fails to resolve metadata field correctly

### ISSUE THREAD

**Issue by antoinebardoz** [conan-io/conan#15650]
### Environment details

* Operating System+version: Ubuntu 22.04
* Compiler+version: N/A
* Conan version: 2.0.17
* Python version: 3.10.8

### Steps to reproduce

1. Have a package with a version number followed by a "metadata (i.e. preceded by +)" field containing a number. For example: ```1.0.0+2```. NOTE: There cannot be a "larger" version available, such as ```1.0.0+3``` if you want to reproduce this.
2. in a conanfile, require ```>=``` this version. For example: ```self.requires("package/[>=1.0.0+2]")```
3. Try to resolve dependencies. Conan will fail with: ```ERROR: Package 'package/[>=1.0.0+2]' not resolved: Version range '>=1.0.0+2' from requirement 'package/[>=1.0.0+2]' [etc, etc...] ```

I can verify that this should not be the case by running:
```
>>> from conans.model.version import Version
>>> Version("1.0.0+2") >= Version("1.0.0+2")
True
>>> Version("1.0.0+2") >= Version("1.0.0+3")
False
```

Note also that if I set the metadata number to one LESS than the exisiting version, it will be able to find it! If I set ```self.requires("package/[>=1.0.0+1]")``` it will be able to find version ```1.0.0+2```.

This seems like a bug to me?

FYI: we use this field to hold  "build number" in case we run another build without changing the version. It increments after each build. It's useful to us to be able to specify that we should not use a build which is older than some build number.

**Comment by AbrilRBS** [comment]
Hi @antoinebardoz thanks a lot for the report.
This seems to be an issue with a somewhat recent change to how Conan handles prereleases and version ranges, where it's not expecting build metadata in the range.

I've opened https://github.com/conan-io/conan/pull/15653 to track the fix, but I anticipate that some more investigation is still needed on our part, will keep you posted :)

**Comment by memsharded** [comment]
I am having a look at this.

Some suggestions related to:

> FYI: we use this field to hold "build number" in case we run another build without changing the version. It increments after each build. It's useful to us to be able to specify that we should not use a build which is older than some build number.

Please note that builds of a package should only happen when it is necessary, that means:
- There are changes in the source code and/or conanfile.py recipe, that would cause a new ``recipe_revision``
- There are changes in the configuration (profiles), changing settings/options that produce a new ``package_id``
- There are changes in the dependencies that produce a new ``package_id``.

The last 2 cases are already modeled by ``package_id`` and changing the version number, even with the build metadata would not be recommended. For the first case, it can make sense if you prefer to somehow rely on explicit build-metadata instead of Conan recipe-revisions. But if there are no changes to the source code, then re-building the package with a new build number in metadata could be considered unnecessary and an anti-pattern.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15653 has closed this ticket, it will be released in next Conan 2.2

the behavior will be that once a ``+build`` is added to a limit in a version range, the automatic inclusion or exclusion of pre-releases in the range boundaries will not be activated, as it happens without build metadata:

- ``[>=1.0.0]`` accepts any ``1.0.0-pre.1`` prereleases (if pre-releases are enabled). This is the current behavior.
- ``[>=1.0.0+b.2]`` will not accept any ``1.0.0-pre.1`` prereleases. This is the fixed behavior, should solve this ticket.

**Comment by antoinebardoz** [comment]
Thanks, for the fix!

I will have to take some time to digest this, but I agree that having a build counter is far from ideal (we are planing to get rid of it eventually).

We will look more closely at what you wrote, I think it'll help us fully understand how we can improve our use of semver.
